### PR TITLE
use-case: deeply replace packages

### LINKED ISSUE #4: Document use case: Deeply replacing packages in Nixpkgs, e.g. security patching

### PR DESCRIPTION
**Pull request by alex-ameen**

### CODE REVIEW

**Comment by roberth** [comment]
Can I pass the baton to @nbp?

EDIT: looks decent, but I don't know the review criteria as I'm not directly involved with the working group.

File: use-cases/deep-replace.org
Comment (infinisil) [inline use-cases/deep-replace.org:1]: Could we just use markdown for uniformity? The entire ecosystem is using markdown for everything! :sweat_smile:

File: use-cases/deep-replace.org
Comment (infinisil) [inline use-cases/deep-replace.org]: I'd say let's keep the noise down to just the essence of it:
```suggestion
pkgs.extend (final: prev: {
  foo = final.callPackage ./my-pkgs/foo {};
})
```

File: use-cases/deep-replace.org
Comment (infinisil) [inline use-cases/deep-replace.org]: I don't think `callPackageWith`, `makeScope` and `override` can be used for this.
In addition to `extend`, there's also `overlays`, `appendOverlays` and `.overrideScope'`

File: use-cases/deep-replace.org
Comment (infinisil) [inline use-cases/deep-replace.org]:
```suggestion
1. [[https://github.com/NixOS/nixpkgs/blob/master/lib/customisation.nix][github:NixOS/nixpkgs://lib/customisation.nix]] routines aren't intuitively understood by many users.
```

But I don't think this is really related to the deep replacement use case, most of these functions there don't do deep replacement.

Instead I think you might want to describe how there's too many different ways of doing deep replacement, and which one to use is not obvious.

File: use-cases/deep-replace.org
Comment (infinisil) [inline use-cases/deep-replace.org]: Another problem: The difference between `self`/`super` (or `final`/`prev`) is hard to understand and using the wrong one can lead to infinite recursion or unoverridable packages

File: use-cases/deep-replace.org
Comment (infinisil) [inline use-cases/deep-replace.org]: Is this a problem with deep replacement?

File: use-cases/deep-replace.org
Comment (infinisil) [inline use-cases/deep-replace.org]: Same here, this seems more of a problem with Flakes than deep replacement.

**Review by infinisil — CHANGES_REQUESTED** [review]

File: use-cases/deep-replace.org
Comment (alex-ameen) [reply use-cases/deep-replace.org]: I'll add an example there. What I mean is that if I do this it's anyone's guess how overlays will act:

```nix
pkgs.extend ( final: prev: {
  nodejs = prev.nodejs-14_x;
  nodePackages = prev.nodePackages.extend ( nfinal: nprev: {
    bar = nfinal.callPackage ./my-pkgs/bar {};
  } );
} )
```

That one's a case where you should probably use two overlays applied separately; but I only know that because I messed this type of thing up dozens of times.

File: use-cases/deep-replace.org
Comment (alex-ameen) [reply use-cases/deep-replace.org]: If a fence between two houses is damaged, each neighbor could blame the other.

I'd just argue that flakes are a common use case for Nixpkgs. I don't want to suggest that changes to `flakes` are in scope here; rather I'm suggesting that the ability to compose multiple ad-hoc extensions to Nixpkgs ( conventionally through flakes ) is worth identifying as a use case.

File: use-cases/deep-replace.md
Comment (bew) [inline use-cases/deep-replace.md]:
```suggestion
   parent scopes and child scopes is opaque to users.
```

File: use-cases/deep-replace.org
Comment (infinisil) [reply use-cases/deep-replace.org]: Ah makes sense!

File: use-cases/deep-replace.md
Comment (infinisil) [inline use-cases/deep-replace.md]: Great example!

Comment (infinisil) [reply use-cases/deep-replace.md]: Wait a second, why does this not work? `prev.nodejs` isn't referenced anywhere in this overlay, so the old value shouldn't be used anywhere either, right?

Comment (DavHau) [reply use-cases/deep-replace.md]: I think it's because `prev.nodePackages.extend` is used instead of `final.nodePackages.extend`

Comment (infinisil) [reply use-cases/deep-replace.md]: I just wanted to try this out, but turns out `nodePackages.callPackage` doesn't exist, so this code doesn't work:
```
error: attribute 'callPackage' missing

       at /home/tweagysil/src/nixpkgs/ext.nix:6:15:

            5|       nodePackages = prev.nodePackages.extend ( nfinal: nprev: {
            6|         bar = nfinal.callPackage ./my-pkgs/bar {};
             |               ^
            7|       } );
```

@aakropotkin Can you give an example that can be reproduced?

Comment (alex-ameen) [reply use-cases/deep-replace.md]: yes absolutely

**Review by phaer — APPROVED** [review]
LGTM, except for the pending update of the example :)

File: use-cases/deep-replace.md
Comment (infinisil) [inline use-cases/deep-replace.md]:
```suggestion
truly prevent a user from accomplishing their goal, but we might suffice to
```

**Review by KFearsoff — COMMENTED** [review]
LGTM, but I think a little more detail should be given about the "deep replacing" exactly.

I think providing a convenient way to deep-replace is cool, but I think an even better solution would be making deep-replace the baseline, and shallow replace to be opt-in.

Perhaps an example would help illustrate the point. I imagine an API like this:
- Deep replace (default and easy):
```
let patched-treesitter = pkgs.nvim-treesitter { # pkgs.nvim-treesitter is a module in this context
  src = { # submodule that is merged on top of defaults
    version = "1.2.4"; # instead of 1.2.3 in this example
    sha256 = "<sha256>";
  };
};
nvim-treesitter.enable = true;
nvim-treesitter.package = patched-treesitter;
nvim-treesitter-terraform.enable = true; # will use patched treesitter
```
- Shallow replace:
```
let patched-treesitter = pkgs.nvim-treesitter { # pkgs.nvim-treesitter is a module in this context
  src = { # submodule that is merged on top of defaults
    version = "1.2.4"; # instead of 1.2.3 in this example
    sha256 = "<sha256>";
  };
};
nvim-treesitter.enable = true;
nvim-treesitter-terraform.enable = true;
nvim-treesitter-terraform.package.inputs = { # attrset of modules
  nvim-treesitter = treesitter-patched; # replace input module with our module
};
```

The main challenge here would be that modules are mutable, and we need to create a deep clone of a module to customize it. We also have to be wary of infinite recursion. But I do feel that "Deep replace" should be the expected behavior when you override `hello.package` option, and "Shallow replace" should be the expected behavior when you override `hello.inputs.glibc`.

EDIT: I think that API would be quite possible if a "package module" is a function like this: `overrideArgs ? {}: if (overrideArgs == {}) then (module) else (overridenModule)`

File: use-cases/deep-replace.md
Comment (DavHau) [reply use-cases/deep-replace.md]: Still, I think the mistake here is using:
```
nodePackages = prev.nodePackages.extend
```
instead of
```
nodePackages = curr.nodePackages.extend
```

If that is correct, then the example doesn't demonstrate a problem with deeply replacing packages, but a bad UX of the override concept, that allows the user to confuse `prev` and `curr`.

**Comment by dblsaiko** [comment]
> security patching

This would cause full rebuilds unlike https://github.com/NixOS/nixpkgs/pull/10851/ wants to do however, from what I can see, right?

File: use-cases/deep-replace.md
Comment (phaer) [inline use-cases/deep-replace.md]:
```suggestion
Nested overlays, "chained" vs. "composed" overlays, as and when to use `prev`
and `final` are common sources of confusion.
```

**Review by phaer — APPROVED** [review]
LGTM!

**Comment by DavHau** [comment]
@aakropotkin In an earlier meeting (which you might have missed) we agreed that the approval of all individual team members is required before a PR can get merged. Therefore this PR has been merged too early.

**Pull request merged**
